iml-gui/crate/src/components/command_modal.rs

```diff
@@ -64,87 +106,114 @@ pub fn update(msg: Msg, model: &mut Model, orders: &mut impl Orders<Msg, GMsg>)
-                        .commands_all
-                        .clone()
-                        .into_iter()
-                        .filter(|x| !is_finished(x))
-                        .collect();
+        Msg::FetchedCommands(commands_data_result) => {
+            match *commands_data_result {
+                Ok(api_list) => {
+                    model.assign_commands(&api_list.objects);
                 }
                 Err(e) => {
-                    error!("Failed to perform fetch_command_status {:#?}", e);
-
+                    error!("Failed to perform FetchedCommands {:#?}", e);
                     orders.skip();
                 }
             }
-            if !model.commands.is_empty() {
-                let (cancel, fut) = sleep_with_handle(POLL_INTERVAL, Msg::Fetch, Msg::Noop);
-
-                model.cancel = Some(cancel);
+            if !is_all_finished(&model.commands) {
+                let (cancel, fut) = sleep_with_handle(POLL_INTERVAL, Msg::FetchTree, Msg::Noop);
+                model.tree_cancel = Some(cancel);
                 orders.perform_cmd(fut);
             }
         }
-        Msg::Open(x) => {
-            model.opens.insert(x);
-        }
-        Msg::Close(x) => {
-            model.opens.remove(&x);
+        Msg::FetchedJobs(jobs_data_result) => match *jobs_data_result {
+            Ok(api_list) => {
+                model.assign_jobs(&api_list.objects);
+            }
+            Err(e) => {
+                error!("Failed to perform FetchJobs {:#?}", e);
+                orders.skip();
+            }
+        },
+        Msg::FetchedSteps(steps_data_result) => match *steps_data_result {
+            Ok(api_list) => {
+                model.assign_steps(&api_list.objects);
+            }
+            Err(e) => {
+                error!("Failed to perform FetchSteps {:#?}", e);
+                orders.skip();
+            }
+        },
+        Msg::Click(the_id) => {
+            let (select, do_fetch) = interpret_click(&model.select, the_id);
+            model.select = select;
+            if do_fetch {
+                schedule_fetch_tree(model, orders);
+            }
         }
         Msg::Noop => {}
     }
 }
 
-async fn fetch_command_status(command_ids: Vec<u32>) -> Result<Msg, Msg> {
-    // e.g. GET /api/command/?id__in=1&id__in=2&id__in=11&limit=0
-    let mut ids: Vec<_> = command_ids.into_iter().map(|x| ("id__in", x)).collect();
-
-    ids.push(("limit", 0));
-
-    Request::api_query(Command::endpoint_name(), &ids)
-        .expect("Bad query for command")
-        .fetch_json_data(|x| Msg::Fetched(Box::new(x)))
-        .await
-}
-
-const fn is_finished(cmd: &Command) -> bool {
-    cmd.complete
+fn schedule_fetch_tree(model: &mut Model, orders: &mut impl Orders<Msg, GMsg>) {
+    match &model.select {
+        Select::None => {
+            // the user has all the commands dropdowns closed
+            let cmd_ids = extract_sorted_keys(&model.commands);
```

Comment:
*Side note*: we may put an additional ordered array `commands_order: Vec<u32>` to `Model` and maintain it to be consistent with `commands`. In this case this will complicate the implementation even more, consume more memory but will be a bit faster, since we don't need to rebuld `cmd_ids` every time.
The same is for `jobs` and `steps`.